src/verify/mod.rs

```diff
@@ -0,0 +1,309 @@
+                        TopLevelItem {
+                            func: Func {
+                                n_params: params.len(),
+                            },
+                            source: item,
+                        },
+                    ) {
+                        errors.push(Error {
+                            kind: error::Kind::ItemConflict(conflict.source, item),
+                            context: error::Context::TopLevel,
+                            source: item.node(),
+                        });
+                    }
+                }
+            }
+        }
+
+        (out, errors)
+    }
+
+    fn check_item(&'a self, item: &'b ast::ItemKind<'b>) -> Vec<Error<'b>> {
+        match item {
+            ast::ItemKind::FnDecl { name: _, params, body } => Scope::check_fn(self, params, body),
+        }
+    }
+
+    fn check_items(&'a self, items: &'b [ast::Item<'b>]) -> Vec<Error<'b>> {
+        let mut errors = Vec::new();
+        for item in items {
+            errors.extend(self.check_item(&item.kind));
+        }
+        errors
+    }
+}
+
+impl<'a, 'b: 'a> Scope<'a> {
+    /// Creates a new scope containing the given item with the given parent.
+    /// The TopLevelScope is inherited from `parent`.
+    fn new(item: ScopeItem<'a>, parent: &'a Scope<'a>) -> Scope<'a> {
+        Scope {
+            item: Some(item),
+            parent: Some(parent),
+            top_level: parent.top_level,
+        }
+    }
+    /// Creates a new empty scope - with no item and no parent.
+    fn empty(top_level: &'a TopLevelScope<'a>) -> Scope<'a> {
+        Scope {
+            item: None,
+            parent: None,
+            top_level,
+        }
+    }
+
+    /// Creates a new scope, a child of `self`, containing `item`.
+    fn child(&'a self, item: ScopeItem<'a>) -> Scope<'a> {
+        Self::new(item, self)
+    }
+
+    /// Given a TopLevelScope, check_fn checks the body of the function and returns any errors.
+    fn check_fn(
+        top_level: &'a TopLevelScope<'a>,
+        params: &'b ast::FnParams<'b>,
+        block: &'b ast::Block<'b>,
+    ) -> Vec<Error<'b>> {
+        // Storage for later
+        let empty;
+        let mut scopes;
+
+        let fn_scope = if params.is_empty() {
+            empty = Self::empty(top_level);
+            &empty
+        } else {
+            // We'll now create a scope for each parameter.
+            // It's a shame we can't do this on the stack :(
+            scopes = Vec::with_capacity(params.len());
+            for param in params {
+                scopes.push(Scope {
+                    item: Some(ScopeItem {
+                        name: param.name,
+                        variable: Variable {},
+                        source: Some(param.node()),
+                    }),
+                    // parent will be filled in later
+                    parent: None,
+                    top_level,
+                });
+            }
```

Comment:
This can probably be more cleanly written with `map` - something like:
```rust
scopes = params.iter().map(|param| {
    Scope {
        item: Some(ScopeItem {
            name: param.name,
            variable: Variable {}
            source: Some(param.node()),
        }),
        // parent will be filled in later
        parent: None,
        top_level,
    }
}).collect();
```